graphite-ch-optimizer.go

```diff
@@ -0,0 +1,380 @@
+	})
+	viper.SetDefault("daemon", map[string]interface{}{
+		"one-shot":      false,
+		"loop-interval": time.Duration(1) * time.Hour,
+		"dry-run":       false,
+	})
+	viper.SetDefault("logging", map[string]interface{}{
+		"output":    "-",
+		"log-level": "info",
+	})
+}
+
+func processFlags() error {
+	// Parse command line arguments in differend flag groups
+	pflag.CommandLine.SortFlags = false
+	customConfig := pflag.StringP("config", "c", "", "Filename of the custom config. CLI arguments override it")
+	pflag.Bool("print-defaults", false, "Print default config values and exit")
+
+	// ClickHouse set
+	fc := pflag.NewFlagSet("clickhouse", 0)
+	fc.StringP("server-dsn", "s", viper.GetString("clickhouse.server-dsn"), "DSN to connect to ClickHouse server")
+	fc.Duration("optimize-interval", viper.GetDuration("clickhouse.optimize-interval"), "The active partitions won't be optimized more than once per this interval, seconds")
+	// Daemon set
+	fd := pflag.NewFlagSet("daemon", 0)
+	fd.Bool("one-shot", viper.GetBool("daemon.one-shot"), "Program will make only one optimization instead of working in the loop (true if dry-run)")
+	fd.Duration("loop-interval", viper.GetDuration("daemon.loop-interval"), "Daemon will check if there partitions to merge once per this interval, seconds")
+	fd.BoolP("dry-run", "n", viper.GetBool("daemon.dry-run"), "Will print how many partitions would be merged without actions")
+	// Logging set
+	fl := pflag.NewFlagSet("logging", 0)
+	fl.String("output", viper.GetString("logging.output"), "The logs file. '-' is accepted as STDOUT")
+	fl.String("log-level", viper.GetString("logging.level"), "Valid options are: panic, fatal, error, warn, warning, info, debug, trace")
+
+	pflag.CommandLine.AddFlagSet(fc)
+	pflag.CommandLine.AddFlagSet(fd)
+	pflag.CommandLine.AddFlagSet(fl)
+
+	pflag.Parse()
+	// We must read config files before the setting of the config config to flags' values
+	err := readConfigFile(customConfig)
+	if err != nil {
+		return err
+	}
+
+	// Parse flag groups into viper config
+	fc.VisitAll(func(f *pflag.Flag) {
+		viper.BindPFlag("clickhouse."+f.Name, f)
+	})
+	fd.VisitAll(func(f *pflag.Flag) {
+		viper.BindPFlag("daemon."+f.Name, f)
+	})
+	fl.VisitAll(func(f *pflag.Flag) {
+		viper.BindPFlag("logging."+f.Name, f)
+	})
+
+	// If it's dry run, then it should be done only once
+	if viper.GetBool("daemon.dry-run") {
+		viper.Set("daemon.one-shot", true)
+	}
+
+	return nil
+}
+
+// readConfigFile set file as the config name if it's not empty and reads the config from Viper.configPaths
+func readConfigFile(file *string) error {
+	var cfgNotFound viper.ConfigFileNotFoundError
+	viper.SetConfigFile(*file)
+	err := viper.ReadInConfig()
+	if err != nil {
+		if errors.As(err, &cfgNotFound) {
+			logrus.Debug("No config files were found, use defaults and flags")
+			return nil
+		}
+		return fmt.Errorf("Failed to read viper config: %w", err)
+	}
+	return nil
+}
+
+func getConfig() *Config {
```

Comment:
returning a pointer is not necessary. remove it

Reply:
possible, yes